### PR TITLE
RavenDB-14963 - Adjustments

### PR DESCRIPTION
**Pull request by shaharhikri**
### Issue link

Extention for the last revisions view: https://github.com/ravendb/ravendb/pull/19311
**Server Side:**
* Add collection to the json output
* Rename public RevertRevisionsByIdOperation(string id, string cv) -> public RevertRevisionsByIdOperation(string id, string changeVector)
* enable to filter by 3 options, [all, deleted, not deleted] also per collection
* add ability to filter revision ids by prefix
* pagesize=0 shouldn't return a result
* revision collection EP should return the same json format as the CollectionsHandlerProcessorForGetCollectionStats EP
**Studio Side:**
* Add collection, flags column
* Add deleted revision marker
* Click on a revision, should go to the revision itself (not the document)
* Show the count of revisions in the menu
* Merge the Revision Bin and All Revisions views
* Unresolved studio revisions view mode (talk to @Pawel Pekrol before proceeding)

Selecting the All Revision will transform the view to show the count and collections of revisions
Filter by collection (button to show revisions in the collection view)
Let's finish the server side first and then handle the studio side.

### Additional description

...Include details of the change made in this Pull Request or additional notes for the solution. Anything that can be useful for reviewers of this PR...

### Type of change

- [ ] Bug fix
- [ ] Regression bug fix
- [ ] Optimization
- [x] New feature

### How risky is the change?

- [x] Low 
- [ ] Moderate 
- [ ] High
- [ ] Not relevant

### Backward compatibility

- [x] Non breaking change
- [ ] Ensured. Please explain how has it been implemented?
- [ ] Breaking change
- [ ] Not relevant

### Is it platform specific issue?

- [ ] Yes. Please list the affected platforms.
- [x] No

### Documentation update

- [x] This change requires a documentation update. Please mark the issue on YouTrack using `Documentation Required` tag.
- [ ] No documentation update is needed 

### Testing by Contributor

- [x] Tests have been added that prove the fix is effective or that the feature works
- [ ] Internal classes added to the test class (e.g. entity or index definition classes) have the lowest possible access modifier (preferable `private`) 
- [ ] It has been verified by manual testing

### Testing by RavenDB QA team

- [ ] This change requires a special QA testing due to possible performance or resources usage implications (CPU, memory, IO). Please mark the issue on YouTrack using `QA Required` tag.
- [x] No special testing by RavenDB QA team is needed

### Is there any existing behavior change of other features due to this change?

- [ ] Yes. Please list the affected features/subsystems and provide appropriate explanation
- [x] No

### UI work

- [x] It requires further work in the Studio. Please mark the issue on YouTrack using `Studio Required` tag.
- [ ] No UI work is needed

### CODE REVIEW

**Comment by shaharhikri** [comment]
test this please

File: src/Raven.Server/Documents/Revisions/RevisionsStorage.cs [hunk 1]
```diff
@@ -1174,31 +1173,127 @@ private long GetConflictRevisionsCount(
             return conflictCount;
         }
 
-        public IEnumerable<Document> GetRevisionsInReverseEtagOrder(DocumentsOperationContext context, int skip, int take)
+        public enum RevisionsType
```
Comment (karmeli87) [inline src/Raven.Server/Documents/Revisions/RevisionsStorage.cs]: we need a better name here.
maybe `FilterRevisionsOption`?

File: src/Raven.Server/Documents/Revisions/RevisionsStorage.cs [hunk 2]
```diff
@@ -1174,31 +1173,127 @@ private long GetConflictRevisionsCount(
...
+
+            using var _ = GetEtagAsSlice(context, 0, out var startSlice);
+
+            switch (type)
+            {
+                case RevisionsType.All:
+                    revisions = table.SeekBackwardFromLast(RevisionsSchema.FixedSizeIndexes[AllRevisionsEtagsSlice], skip);
+                    break;
+                case RevisionsType.NotDeleted:
+                    revisions = table.SeekBackwardFrom(RevisionsSchema.Indexes[DeleteRevisionEtagSlice], null, startSlice, skip).Select(seekResult => seekResult.Result);
```
Comment (karmeli87) [inline src/Raven.Server/Documents/Revisions/RevisionsStorage.cs]: please use named parameters

File: src/Raven.Server/Documents/Revisions/RevisionsStorage.cs [hunk 3]
```diff
@@ -1174,31 +1173,127 @@ private long GetConflictRevisionsCount(
...
+                    break;
+                case RevisionsType.Deleted:
+                    revisions = table.SeekBackwardFrom(RevisionsSchema.Indexes[DeleteRevisionEtagSlice], null, Slices.AfterAllKeys, skip).Select(seekResult => seekResult.Result);
```
Comment (karmeli87) [inline src/Raven.Server/Documents/Revisions/RevisionsStorage.cs]: please use named parameters

File: src/Raven.Server/Documents/Revisions/RevisionsStorage.cs [hunk 4]
```diff
@@ -1174,31 +1173,127 @@ private long GetConflictRevisionsCount(
...
+                {
+                    case RevisionsType.All:
+                        break;
+                    case RevisionsType.NotDeleted:
+                        if (revision.Flags.Contain(DocumentFlags.DeleteRevision))
+                            continue;
+                        break;
+                    case RevisionsType.Deleted:
+                        if (revision.Flags.Contain(DocumentFlags.DeleteRevision) == false)
+                            continue;
```
Comment (karmeli87) [inline src/Raven.Server/Documents/Revisions/RevisionsStorage.cs]: if we skip, we need to dispose it

File: src/Raven.Server/Web/Studio/Processors/AbstractStudioCollectionsHandlerProcessorForPreviewRevisions.cs [hunk 5]
```diff
@@ -20,9 +22,18 @@ internal abstract class AbstractStudioCollectionsHandlerProcessorForPreviewRevis
 
     protected string Collection;
 
+    protected RevisionsStorage.RevisionsType Type;
+
     protected AbstractStudioCollectionsHandlerProcessorForPreviewRevisions([NotNull] TRequestHandler requestHandler) : base(requestHandler)
     {
         ContextPool = RequestHandler.ContextPool;
+        Collection = RequestHandler.GetStringQueryString("collection", required: false);
+        var type = RequestHandler.GetStringQueryString("type", required: false) ?? "all";
```
Comment (karmeli87) [inline src/Raven.Server/Web/Studio/Processors/AbstractStudioCollectionsHandlerProcessorForPreviewRevisions.cs]: type -> filterOption
and the default should be typed and not a string value

File: src/Raven.Server/Web/Studio/StudioCollectionsHandlerProcessorForPreviewRevisions.cs [hunk 6]
```diff
@@ -18,18 +20,38 @@ internal sealed class StudioCollectionsHandlerProcessorForPreviewRevisions : Abs
...
+            case RevisionsStorage.RevisionsType.All:
+                _totalResults = string.IsNullOrEmpty(Collection)
+                    ? RequestHandler.Database.DocumentsStorage.RevisionsStorage.GetNumberOfRevisionDocuments(context)
+                    : RequestHandler.Database.DocumentsStorage.RevisionsStorage.GetNumberOfRevisionDocumentsForCollection(context, Collection);
+                break;
+
+            case RevisionsStorage.RevisionsType.NotDeleted:
+                _totalResults = string.IsNullOrEmpty(Collection)
+                    ? RequestHandler.Database.DocumentsStorage.RevisionsStorage.GetNumberOfNonDeletedRevisions(context)
+                    : -1; // Not available for specific collection
```
Comment (karmeli87) [inline src/Raven.Server/Web/Studio/StudioCollectionsHandlerProcessorForPreviewRevisions.cs]: make this a const and use it in all places

File: src/Raven.Server/Web/Studio/StudioCollectionsHandlerProcessorForPreviewRevisions.cs [hunk 7]
```diff
@@ -40,48 +62,60 @@ protected override IDisposable OpenReadTransaction(DocumentsOperationContext con
     protected override Task WriteItemsAsync(DocumentsOperationContext context, AsyncBlittableJsonTextWriter writer)
     {
         writer.WriteStartArray();
+        if (_totalResults != 0)
```
Comment (karmeli87) [inline src/Raven.Server/Web/Studio/StudioCollectionsHandlerProcessorForPreviewRevisions.cs:71]: add a comment here that you use `_totalResults` for 3 cases
0 - no results
-1 - unknown number of results
else - we have results

File: test/SlowTests/Issues/RavenDB-14963.cs [hunk 8]
```diff
@@ -31,8 +29,8 @@ public RavenDB_14963(ITestOutputHelper output) : base(output)
     }
 
     [RavenTheory(RavenTestCategory.Revisions | RavenTestCategory.Voron)]
-    [RavenData(DatabaseMode = RavenDatabaseMode.All, Data = new object[] { true })]
-    [RavenData(DatabaseMode = RavenDatabaseMode.All, Data = new object[] { false })]
+    [RavenData(DatabaseMode = RavenDatabaseMode.Single, Data = new object[] { true })]
+    [RavenData(DatabaseMode = RavenDatabaseMode.Single, Data = new object[] { false })]
```
Comment (karmeli87) [inline test/SlowTests/Issues/RavenDB-14963.cs]: why single?

**Review by karmeli87 — APPROVED** [review]
In general LGTM, left few cosmetic comments

File: test/SlowTests/Issues/RavenDB-14963.cs (hunk 8 above)
Comment (shaharhikri) [reply test/SlowTests/Issues/RavenDB-14963.cs]: Done

File: src/Raven.Server/Web/Studio/StudioCollectionsHandlerProcessorForPreviewRevisions.cs (hunk 7 above)
Comment (shaharhikri) [reply src/Raven.Server/Web/Studio/StudioCollectionsHandlerProcessorForPreviewRevisions.cs:71]: Done

File: src/Raven.Server/Web/Studio/StudioCollectionsHandlerProcessorForPreviewRevisions.cs (hunk 6 above)
Comment (shaharhikri) [reply src/Raven.Server/Web/Studio/StudioCollectionsHandlerProcessorForPreviewRevisions.cs]: Done

File: src/Raven.Server/Web/Studio/Processors/AbstractStudioCollectionsHandlerProcessorForPreviewRevisions.cs (hunk 5 above)
Comment (shaharhikri) [reply src/Raven.Server/Web/Studio/Processors/AbstractStudioCollectionsHandlerProcessorForPreviewRevisions.cs]: Done

File: src/Raven.Server/Documents/Revisions/RevisionsStorage.cs (hunk 4 above)
Comment (shaharhikri) [reply src/Raven.Server/Documents/Revisions/RevisionsStorage.cs]: Done

File: src/Raven.Server/Documents/Revisions/RevisionsStorage.cs (hunk 3 above)
Comment (shaharhikri) [reply src/Raven.Server/Documents/Revisions/RevisionsStorage.cs]: Done

File: src/Raven.Server/Documents/Revisions/RevisionsStorage.cs (hunk 2 above)
Comment (shaharhikri) [reply src/Raven.Server/Documents/Revisions/RevisionsStorage.cs]: Done

File: src/Raven.Server/Documents/Revisions/RevisionsStorage.cs (hunk 1 above)
Comment (shaharhikri) [reply src/Raven.Server/Documents/Revisions/RevisionsStorage.cs]: Done

File: src/Raven.Server/Documents/Revisions/RevisionsStorage.cs [hunk 9]
```diff
@@ -1174,31 +1174,130 @@ private long GetConflictRevisionsCount(
...
+
+            using var _ = GetEtagAsSlice(context, 0, out var startSlice);
+
+            switch (type)
+            {
+                case FilterRevisionsOption.All:
+                    revisions = table.SeekBackwardFromLast(RevisionsSchema.FixedSizeIndexes[AllRevisionsEtagsSlice], skip);
+                    break;
+                case FilterRevisionsOption.NotDeleted:
+                    revisions = table.SeekBackwardFrom(RevisionsSchema.Indexes[DeleteRevisionEtagSlice], prefix: null, last: startSlice, skip).Select(seekResult => seekResult.Result);
```
Comment (ppekrol) [inline src/Raven.Server/Documents/Revisions/RevisionsStorage.cs]: avoid LINQ

File: src/Raven.Server/Documents/Revisions/RevisionsStorage.cs [hunk 10]
```diff
@@ -1174,31 +1174,130 @@ private long GetConflictRevisionsCount(
...
+                    break;
+                case FilterRevisionsOption.Deleted:
+                    revisions = table.SeekBackwardFrom(RevisionsSchema.Indexes[DeleteRevisionEtagSlice], prefix: null, last: Slices.AfterAllKeys, skip).Select(seekResult => seekResult.Result);
```
Comment (ppekrol) [inline src/Raven.Server/Documents/Revisions/RevisionsStorage.cs]: avoid LINQ

File: src/Raven.Server/Web/Studio/Processors/AbstractStudioCollectionsHandlerProcessorForPreviewRevisions.cs [hunk 11]
```diff
@@ -20,9 +22,24 @@ internal abstract class AbstractStudioCollectionsHandlerProcessorForPreviewRevis
 
     protected string Collection;
 
+    protected RevisionsStorage.FilterRevisionsOption Type;
+
     protected AbstractStudioCollectionsHandlerProcessorForPreviewRevisions([NotNull] TRequestHandler requestHandler) : base(requestHandler)
     {
         ContextPool = RequestHandler.ContextPool;
+        Collection = RequestHandler.GetStringQueryString("collection", required: false);
+        var type = RequestHandler.GetStringQueryString("filterOption", required: false);
```
Comment (ppekrol) [inline src/Raven.Server/Web/Studio/Processors/AbstractStudioCollectionsHandlerProcessorForPreviewRevisions.cs]: filterOption => type

File: src/Raven.Server/Web/Studio/Sharding/Processors/ShardedStudioCollectionsHandlerProcessorForPreviewRevisions.cs [hunk 12]
```diff
@@ -125,6 +130,9 @@ protected override async Task InitializeAsync(TransactionOperationContext contex
 
     protected override async ValueTask<long> GetTotalCountAsync()
     {
+        if (Type != RevisionsStorage.FilterRevisionsOption.All && string.IsNullOrEmpty(Collection) == false) // In that case the shards results won't contain 'TotalResults' field
+            return -1;
```
Comment (ppekrol) [inline src/Raven.Server/Web/Studio/Sharding/Processors/ShardedStudioCollectionsHandlerProcessorForPreviewRevisions.cs]: can we have a const somewhere?

File: src/Raven.Server/Web/Studio/Processors/AbstractStudioCollectionsHandlerProcessorForPreviewRevisions.cs (hunk 11 above)
Comment (ppekrol) [reply src/Raven.Server/Web/Studio/Processors/AbstractStudioCollectionsHandlerProcessorForPreviewRevisions.cs]: FilterOption and the enum name is misleading. E.g. 'filter=All' - what is your expectation? For me this is 'filter our all the results' so... 0 results should be returned. Please think about a better naming.

Comment (karmeli87) [reply src/Raven.Server/Web/Studio/Processors/AbstractStudioCollectionsHandlerProcessorForPreviewRevisions.cs]: so maybe `SelectionOption`/`RevisionSelectionOption` ?

File: src/Raven.Server/Documents/Revisions/RevisionsStorage.cs (hunk 9 above)
Comment (shaharhikri) [reply src/Raven.Server/Documents/Revisions/RevisionsStorage.cs]: k

File: src/Raven.Server/Documents/Revisions/RevisionsStorage.cs (hunk 10 above)
Comment (shaharhikri) [reply src/Raven.Server/Documents/Revisions/RevisionsStorage.cs]: k

File: src/Raven.Server/Web/Studio/Processors/AbstractStudioCollectionsHandlerProcessorForPreviewRevisions.cs (hunk 11 above)
Comment (shaharhikri) [reply src/Raven.Server/Web/Studio/Processors/AbstractStudioCollectionsHandlerProcessorForPreviewRevisions.cs]: k

File: src/Raven.Server/Web/Studio/Sharding/Processors/ShardedStudioCollectionsHandlerProcessorForPreviewRevisions.cs (hunk 12 above)
Comment (shaharhikri) [reply src/Raven.Server/Web/Studio/Sharding/Processors/ShardedStudioCollectionsHandlerProcessorForPreviewRevisions.cs]: k

**Pull request merged**